kubernetes/kubernetes.go

```diff
@@ -0,0 +1,223 @@
+		sourcePort = i32toPort(s.Port)
+		if i32toPort(s.TargetPort.IntVal) > 0 {
+			targetPort = i32toPort(s.TargetPort.IntVal)
+		} else {
+			targetPort = sourcePort
+		}
+
+		if sourcePort > 0 && targetPort > 0 {
+			out[PortMatch{
+				Object: Object{
+					Name:      service.Name,
+					Namespace: service.Namespace,
+				},
+				Port: sourcePort,
+			}] = Service{
+				Port:       uint16(s.Port),
+				TargetPort: targetPort,
+			}
+		}
+	}
+
+	return out
+}
+
+func getEndpoints(client v12.EndpointsInterface) (map[Object][]Endpoint, error) {
+	endpoints, err := client.List(v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	out := make(map[Object][]Endpoint)
+	for _, e := range endpoints.Items {
+		out[Object{Name: e.Name, Namespace: e.Namespace}] = mapEndpoint(e)
+	}
+
+	return out, nil
+}
+
+func getIngresses(client v1beta1.IngressInterface) (map[HostMatch]Ingress, error) {
+	ingresses, err := client.List(v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	out := make(map[HostMatch]Ingress)
+	for _, i := range ingresses.Items {
+		for host, backend := range mapIngress(i) {
+			out[HostMatch{
+				Object:   Object{Name: backend.Name, Namespace: i.Namespace},
+				HostName: host,
+			}] = backend
+		}
+	}
+
+	return out, nil
+}
+
+func mapEndpoint(in v13.Endpoints) []Endpoint {
+	out := make([]Endpoint, 0)
+	for _, s := range in.Subsets {
+		for _, p := range s.Ports {
+			if p.Protocol == v13.ProtocolTCP && i32toPort(p.Port) > 0 {
+				for _, a := range s.Addresses {
+					out = append(out, Endpoint{
+						Address: a.IP,
+						Port:    uint16(p.Port),
+					})
+				}
+			}
+		}
+	}
+
+	return out
+}
+
+func mapIngress(in v1beta12.Ingress) map[string]Ingress {
+	out := make(map[string]Ingress)
+
+	for _, r := range in.Spec.Rules {
+
+		//find suitable path target (we only support / for now)
+		var backend *Ingress
+		for _, p := range r.HTTP.Paths {
+			if p.Path == "" || p.Path == "/" {
+				backend = mapBackend(in.Namespace, p.Backend)
+			}
+		}
+
+		if r.Host != "" && backend != nil {
+			out[r.Host] = *backend
+		} else {
+			log.Debug("Ignoring ingress rule with no hostname, suitable backend, catch-all path or rule for /",
+				zap.String("name", in.Name),
+				zap.String("namespace", in.Namespace),
+				zap.String("host", r.Host))
+		}
+	}
+
+	return out
+}
+
+func mapBackend(namespace string, backend v1beta12.IngressBackend) *Ingress {
+
+	port := backend.ServicePort.IntValue()
+	if toPort(port) == 0 {
+		log.Warn("Dropping ingress backend with invalid port (hint: port names not supported)",
+			zap.String("name", backend.ServiceName),
+			zap.String("namespace", namespace),
+			zap.String("port", backend.ServicePort.String()))
+
+		return nil
+	}
+	return &Ingress{
+		Name:   backend.ServiceName,
+		Port:   uint16(port),
+		Scheme: "http",
+	}
+}
+
+func toPort(port int) uint16 {
+	if port > 0 && port < math.MaxUint16 {
+		return uint16(port)
+	} else {
+		return 0
+	}
+}
+
+func i32toPort(port int32) uint16 {
```

Comment:
Perhaps yes but.. what about the case where port is 0 ? I probably have to check for that anyway somewhere (yes I know that 0 is technically a legal port :P) .. So that would make my outer check look like `if err != nil && port > 0` rather than just `port > 0`. The way I see it, it just "moving the uglyness to somewhere else" :P

Reply:
Ok right.. I could of course just decide that it's an error too, if port == 0.

Reply:
port == 0 usually means "pick a random port", right? So if you get 0 as port through the API, you're screwed either way. Or does k8s do something different with that?